dialogue-target/src/main/java/com/palantir/dialogue/Request.java

```diff
@@ -119,40 +123,32 @@ public static Builder builder() {
 
     @NotThreadSafe
     public static final class Builder {
-        private static final ImmutableListMultimap<String, String> EMPTY =
-                new ImmutableListMultimap.Builder<String, String>().build();
 
-        @Nullable
-        private ListMultimap<String, String> headerParams;
+        @SuppressWarnings("UnnecessaryLambda") // Avoid unnecessary allocation
+        private static final com.google.common.base.Supplier<List<String>> MAP_VALUE_FACTORY = () -> new ArrayList<>(1);
 
-        @Nullable
-        private ImmutableListMultimap.Builder<String, String> queryParams;
+        private static final int MUTABLE_HEADERS = 1;
+        private static final int MUTABLE_QUERY = 1 << 1;
+        private static final int MUTABLE_PATH = 1 << 2;
 
-        @Nullable
-        private ImmutableMap.Builder<String, String> pathParams;
+        private ListMultimap<String, String> headerParams = ImmutableListMultimap.of();
 
-        private Optional<RequestBody> body = Optional.empty();
+        private ListMultimap<String, String> queryParams = ImmutableListMultimap.of();
 
-        // To optimize the case where a builder doesn't end up modifying headers/queryparams/pathparams, we may store
-        // references to another Request's internal objects. If we need to do a mutation, then we'll copy the elements
-        // and null these out.
-        @Nullable
-        private ListMultimap<String, String> existingUnmodifiableHeaderParams;
+        private Map<String, String> pathParams = ImmutableMap.of();
 
-        @Nullable
-        private ImmutableListMultimap<String, String> existingQueryParams;
+        private Optional<RequestBody> body = Optional.empty();
 
-        @Nullable
-        private ImmutableMap<String, String> existingPathParams;
+        private int mutableCollections = 0;
```

Comment:
would kinda like to include 'bitset' or 'bitflags' in the name here so that people don't think this is a count of mutable collections...